pkg/services/ngalert/store/alertmanager.go

```diff
@@ -33,6 +35,23 @@ func (st *DBstore) GetLatestAlertmanagerConfiguration(query *models.GetLatestAle
 	})
 }
 
+// GetAllLatestAlertmanagerConfiguration returns the latest configuration of every organization
+func (st *DBstore) GetAllLatestAlertmanagerConfiguration() ([]*models.AlertConfiguration, error) {
+	var result []*models.AlertConfiguration
+	err := st.SQLStore.WithDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		condition := builder.In("id", builder.Select("MAX(id)").From("alert_configuration").GroupBy("org_id"))
```

Comment:
Just to check that we have an index on `alert_configuration` over `org_id, id`?

Reply:
I think the existing index that includes `org_id` should be sufficient. To confirm my guess I tested against a database (MySql) with 1kk records and the index did not improve the query execution time

Reply:
I do not understand what filtering by org_id will give us here. The query is supposed to return the latest configuration for every organization.
```sql
SELECT A1.* FROM alert_configuration as A1 
WHERE A1.Id IN (SELECT MAX(A2.id) FROM alert_configuration AS A2 GROUP BY A2.org_id)
```
This query produces a plan (MySql 5.6), which seems to be optimal from my point of view.

| id | select\_type | table | type | possible\_keys | key | key\_len | ref | rows | Extra |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | PRIMARY | alert\_configuration | ALL | NULL | NULL | NULL | NULL | 857142 | Using where |
| 2 | SUBQUERY | alert\_configuration | range | IDX\_alert\_configuration\_org\_id,alert\_configuration\_org\_id\_id\_index | alert\_configuration\_org\_id\_id\_index | 8 | NULL | 21979 | Using index for group-by |

Reply:
What happens in the case where an alert configuration exist but an organization has been removed? As far as I know, there's nothing in place yet (but feel free to pick it up as a requisite to this https://github.com/grafana/grafana/issues/39218) that removes the configuration for an org that has been removed. 

Technically, there's nothing wrong with the implementation but you'll bring _all_ existing configuration regardless if the org is suppose to have an alertmanager or not. Or am I missing something?

Reply:
You are correct, we do not have a foreign key between alert_configuration and org. However,
1. I would consider a case when org does not exist to be an exceptional case - data integrity problem. We should fix the problem (ie add FK) rather than try to handle it.
2. Adding the filter by org decreases performance of the query. It is executed 3x times slower on the data set I did benchmarks testing (see the first comment).

So, I think that pulling an extra config, which should not happen too often, is better than filtering by orgs.

Reply:
> I would consider a case when org does not exist to be an exceptional case - data integrity problem. We should fix the problem (ie add FK) rather than try to handle it.

We can agree that this is a problem that needs fixing. I have added a comment to https://github.com/grafana/grafana/issues/39218 to make sure we fix this.

> So, I think that pulling an extra config, which should not happen too often, is better than filtering by orgs.

sgtm.

Reply:
More details about the query:
```sql
SELECT SQL_NO_CACHE * FROM alert_configuration 
WHERE id IN (SELECT MAX(id) FROM alert_configuration AS A2 WHERE EXISTS (SELECT 1 FROM org AS O WHERE O.id = A2.org_id) GROUP BY org_id)
```
that produces plan

| id | select\_type | table | type | possible\_keys | key | key\_len | ref | rows | Extra |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | PRIMARY | alert\_configuration | ALL | NULL | NULL | NULL | NULL | 857142 | Using where |
| 2 | SUBQUERY | A2 | index | IDX\_alert\_configuration\_org\_id | IDX\_alert\_configuration\_org\_id | 8 | NULL | 857142 | Using where; Using index |
| 3 | DEPENDENT SUBQUERY | O | eq\_ref | PRIMARY | PRIMARY | 8 | grafana.A2.org\_id | 1 | Using index |

it's exec time (no caching) 1s 404ms  vs original query exec time (no caching) 685ms


I rewrote it to look like 
```sql 
SELECT A1.* FROM alert_configuration AS A1
WHERE A1.id IN (SELECT MAX(id) FROM alert_configuration AS A2 GROUP BY org_id)
  AND EXISTS (SELECT 1 FROM org AS O WHERE O.id = A1.org_id);
```

and it did improve the exec time to match the original request query time (no chacing) 710ms


| id | select\_type | table | type | possible\_keys | key | key\_len | ref | rows | Extra |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | PRIMARY | A1 | ALL | NULL | NULL | NULL | NULL | 857142 | Using where |
| 3 | DEPENDENT SUBQUERY | O | eq\_ref | PRIMARY | PRIMARY | 8 | grafana.A1.org\_id | 1 | Using index |
| 2 | SUBQUERY | A2 | range | IDX\_alert\_configuration\_org\_id | IDX\_alert\_configuration\_org\_id | 8 | NULL | 23810 | Using index for group-by |


So, `2` argument is not relevant. However, working around potential data inconsistency rather than properly fixing it does not feel right.